### PR TITLE
[REVIEW] Incorporate Date/Nagi implementation of Hungarian Algorithm

### PR DESCRIPTION
**Pull request by ChuckHastings**
This PR incorporates the Date/Nagi implementation of the Hungarian Algorithm, an exact solution to the Linear Assignment Problem.  The implementation actually solves the Linear Assignment Problem for a batch of matrices.

I incorporated the Date/Nagi software and then adapted it to fit within the RAPIDS/RAFT coding conventions.

This will be exposed initially in cuGraph, although could potentially be relevant for cuML.  The implementation operates on a square dense bipartite adjacency matrix.

SWIPAT approval as of 11/10/2020.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: 3rdPartyLicense/LICENSE_Date_Nagi [hunk 1]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (dantegd) [inline 3rdPartyLicense/LICENSE_Date_Nagi:1]: Small comment: `thirdparty/LICENSES/...` folder would be my suggestion for the folder containing all third party licenses

**Review by dantegd — CHANGES_REQUESTED** [review]

File: 3rdPartyLicense/LICENSE_Date_Nagi (hunk 1 above)
Comment (ChuckHastings) [reply 3rdPartyLicense/LICENSE_Date_Nagi:1]: Done in next push.

File: cpp/include/raft/lap/lap_kernels.cuh [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+        min = slack;
+      }
+    }
+
+    d_col_duals[spid * N + colid] = min;
+  }
+}
+
+// Kernel for calculating initial assignments.
+//  TODO:  N should be vertex_t
```
Comment (dantegd) [inline cpp/include/raft/lap/lap_kernels.cuh:199]: todo is no longer needed, right?

File: cpp/include/raft/lap/lap_kernels.cuh [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+      }
+    }
+
+    d_row_duals[spid * N + rowid] = min;
+  }
+}
+
+// Kernel for reducing the column by subtracting column minimum from each column element.
+//  FIXME:  Once cuda 10.2 is the standard should replace passing infinity
+//          here with using cuda::std::numeric_limits<weight_t>::max()
```
Comment (dantegd) [inline cpp/include/raft/lap/lap_kernels.cuh:171]: might be a good idea to open a github issue to track

File: cpp/include/raft/lap/lap_kernels.cuh [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+    } else {
+      d_row_visited[index] = ACTIVE;
+    }
+  }
+}
+
+// Kernel for populating the predicate matrix for edges in row major format.
+//
+// TODO: N should be of type vertex_t.  Need to clean up things to make
+//       that happen
```
Comment (dantegd) [inline cpp/include/raft/lap/lap_kernels.cuh:256]: comment no longer needed? similar for others about `N` needing to be `verted_t`

File: cpp/include/raft/lap/lap_kernels.cuh [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+      d_addresses[index] = 1;
+    } else {
+      d_predicates[index] = false;
+      d_addresses[index] = 0;
+    }
+  }
+}
+
+// Kernel for scattering the edges based on the scatter addresses.
+//   TODO:  N should be type vertex_t
```
Comment (dantegd) [inline cpp/include/raft/lap/lap_kernels.cuh:279]: same

File: cpp/include/raft/lap/lap_kernels.cuh [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+    }
+    if (rowid == 0) {
+      d_ptrs[spid] = compid;
+      d_ptrs[SP] = M;
+    }
+  }
+}
+
+// Kernel for finding the minimum zero cover.
+//template <typename vertex_t, typename weight_t>
```
Comment (dantegd) [inline cpp/include/raft/lap/lap_kernels.cuh:305]: remove commented line

File: cpp/test/lap/lap.cu [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        for (int k = 0; k < batchsize; k++) {
+          std::cout << j << ":" << i << ":" << k << ":"
+                    << lpx.getPrimalObjectiveValue(k) << ":"
+                    << lpx.getDualObjectiveValue(k) << ":" << total_time
+                    << std::endl;
+        }
+      }
+
+      //			Use getAssignmentVector API to get the optimal row assignments for specified problem id.
+      //			Example is shown below.
```
Comment (dantegd) [inline cpp/test/lap/lap.cu:107]: small question, why is the example here?

**Review by dantegd — CHANGES_REQUESTED** [review]
I just have some very minor comments, though it had been _quite_ some time since I saw the Hungarian algorithm, so another set of eyes might be a good idea.

File: cpp/include/raft/lap/lap_kernels.cuh (hunk 2 above)
Comment (ChuckHastings) [reply cpp/include/raft/lap/lap_kernels.cuh]: Removed

File: cpp/include/raft/lap/lap_kernels.cuh (hunk 4 above)
Comment (ChuckHastings) [reply cpp/include/raft/lap/lap_kernels.cuh]: Removed

File: cpp/include/raft/lap/lap_kernels.cuh (hunk 5 above)
Comment (ChuckHastings) [reply cpp/include/raft/lap/lap_kernels.cuh]: Removed

File: cpp/include/raft/lap/lap_kernels.cuh (hunk 6 above)
Comment (ChuckHastings) [reply cpp/include/raft/lap/lap_kernels.cuh]: Removed

File: cpp/test/lap/lap.cu (hunk 7 above)
Comment (ChuckHastings) [reply cpp/test/lap/lap.cu]: This was code from a third party and I had accidentally left these existing comments in the code.  I have deleted these comments.

File: cpp/include/raft/lap/lap_kernels.cuh (hunk 2 above)
Comment (ChuckHastings) [reply cpp/include/raft/lap/lap_kernels.cuh]: #96 created

File: cpp/include/raft/lap/lap_kernels.cuh (hunk 3 above)
Comment (ChuckHastings) [reply cpp/include/raft/lap/lap_kernels.cuh:171]: Created #96

**Review by dantegd — APPROVED** [review]
Code and changes look good from my perspective. @ChuckHastings not sure if you want an extra set of eyes from someone from the cugraph team?

File: cpp/include/raft/lap/lap.cuh [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    initializeDevice();
+
+    d_vertices_dev.row_assignments = d_row_assignment;
+    d_vertices_dev.col_assignments = d_col_assignment;
+
+    d_costs_ = d_cost_matrix;
+
+    int step = 0;
+
+    while (step != 100) {
```
Comment (BradReesWork) [inline cpp/include/raft/lap/lap.cuh:93]: this seems ripe for an endless loop situation.  seems like there should be a default case statement to catch issues

**Review by BradReesWork — APPROVED** [review]

File: cpp/include/raft/lap/lap.cuh (hunk 8 above)
Comment (ChuckHastings) [reply cpp/include/raft/lap/lap.cuh:93]: The algorithm is guaranteed to converge and eventually return 100 unless there's a bug in implementation.  The return values are all constants and never computed:
* hungarianStep0 always returns 1
* hungarianStep1 can only return a 5 or a 6
* hungarianStep2 can only return a 3 or a 6
* hungarianStep3 can only return a 4 or a 5
* hungarianStep4 always returns 2
* hungarianStep5 always returns 3
* hungarianStep6 always returns 100

**Pull request merged**